crates/threads-xform/src/lib.rs

```diff
@@ -165,7 +170,25 @@ impl Config {
         // should not be called from an agent that cannot block (e.g. the main document thread).
         inject_destroy(module, &tls, &stack, memory)?;
 
-        Ok(())
+        Ok(Some(ThreadCounterAddr(thread_counter_addr)))
+    }
+}
+
+impl ThreadCounterAddr {
+    pub fn wrap_start(self, memory: MemoryId, mut body: InstrSeqBuilder, id: FunctionId) {
+        // This happens after the thread counter was already increased, so "1"
+        // means we are in the first thread.
+        body.i32_const(self.0)
+            .load(memory, LoadKind::I32 { atomic: true }, ATOMIC_MEM_ARG)
```

Comment:
I think this could lead to a race condition - if a second thread starts between when the counter is incremented from 0 to 1 and when we load it here, this could be read as 2 instead of 1 despite this being the first thread, and then the start function won't get run at all.

This'll have to somehow get access to the result of the initial fetch-add done in `inject_start` to avoid the race condition.

Reply:
So the reason why I thought this can not become a race condition is because this is called as part of the `init` function and you can't start a second thread because you don't have access to the WASM instantiation until the `init` function is finished.
I don't consider this brittle either because changing this behavior would be a really big change.

Still, I consider your suggestions an improvement, I will look into it now.

Reply:
So I found two ways to potentially solve this:
1. The injected thread shim returns the thread counter, which we can take later and work with. The problem is that if we change the order of when shims are injected or if we insert a new shim, this will break, and I'm not sure how to fix it then except if make things more complicated and pass the thread counter around.
2. Save the thread counter into a thread local variable. Unfortunately I don't know how to make a thread local variable.

I consider the first solution worse then the current one, so my preference is to keep the current solution as I'm fairly certain it is correct.

Reply:
I went through my assumptions again, I'm certain that this can't become a race condition.
Simply put, to make this a race condition the `init` function has to return the instantiated module without calling the start (`__wbindgen_start`) function.

Reply:
`init` isn't the only way to get a compiled module, though - you can use `WebAssembly.compile` to compile the module in advance, create a `SharedArrayBuffer`, and pass them to two threads which call `init` at the same time.

(I'm not sure what you mean by an 'instantiated module', but `init` takes a compiled `WebAssembly.Module` when initialising another thread.)

Reply:
> `init` isn't the only way to get a compiled module, though - you can use `WebAssembly.compile` to compile the module in advance, create a `SharedArrayBuffer`, and pass them to two threads which call `init` at the same time.

I thought that this isn't supported? For example I took it that your comment here states that: https://github.com/rustwasm/wasm-bindgen/issues/3224#issuecomment-1372920396.

I guess if this is supported, then I have to implement the solution lined out above.
> (I'm not sure what you mean by an 'instantiated module', but `init` takes a compiled `WebAssembly.Module` when initialising another thread.)

Yes, that is what I mean. You can't get the `WebAssembly.Module` without calling the `init` function, unless, like you pointed out, you don't use the `init` function and compile it yourself.

Reply:
See #3234, which addresses this issue.